core/src/main/scala/org/apache/spark/status/api/v1/OneApplicationResource.scala

```diff
@@ -51,6 +52,46 @@ private[v1] class AbstractApplicationResource extends BaseAppResource {
   @Path("executors")
   def executorList(): Seq[ExecutorSummary] = withUI(_.store.executorList(true))
 
+  @GET
+  @Path("executors/{executorId}/threads")
+  def threadDump(@PathParam("executorId") executorId: String): Response = withUI { ui =>
+    uiRoot.executorThreadDumpsNotAvailableError().getOrElse {
+      val safeExecutorId =
+        Option(UIUtils.stripXSS(executorId)).map { executorId =>
+          UIUtils.decodeURLParameter(executorId)
+        }.getOrElse {
+          throw new IllegalArgumentException(s"Missing executorId parameter")
+        }
+
+      def isAllDigits(x: String) = x.forall(Character.isDigit)
+
+      if (executorId != SparkContext.DRIVER_IDENTIFIER && !isAllDigits(executorId)) {
+        Response.serverError()
+          .entity(s"Invalid executorId: neither '${SparkContext.DRIVER_IDENTIFIER}' nor number.")
+          .status(Response.Status.BAD_REQUEST)
+          .build()
+      } else {
+        if (ui.store.asOption(ui.store.executorSummary(executorId)).exists(!_.isActive)) {
+          Response.serverError()
+            .entity("Executor is already dead.")
+            .status(Response.Status.BAD_REQUEST)
+            .build()
+        } else {
+          val maybeStackTraces = ui.sc.flatMap { sc =>
+            sc.getExecutorThreadDump(safeExecutorId)
+          }
+          maybeStackTraces match {
+            case Some(stackTraces) => Response.ok(stackTraces).build()
+            case None => Response.serverError()
+              .entity("No stack traces are available.")
+              .status(Response.Status.NOT_FOUND)
+              .build()
```

Comment:
nit: I think the above code can be simplified as: `ui.sc.flatMap { xxx }.map { xxx }.getOrElse(xxx)`.